addons/website_sale/models/product_feed.py

```diff
@@ -0,0 +1,381 @@
+        """Prepare Google Merchant Center items' fields.
+
+        See Google's (https://support.google.com/merchants/answer/7052112) documentation for more
+        information about each field.
+
+        :return: a dictionary for each product in this recordset.
+        :rtype: list[dict]
+        """
+        products = self._get_feed_products()
+        base_url = self.website_id.get_base_url()
+
+        def format_product_link(url_):
+            if self.pricelist_id:
+                url_ = urlparse(url_)._replace(query=f'pricelist={self.pricelist_id.id}').geturl()
+            return urls.urljoin(
+                base_url, self.env['ir.http']._url_lang(url_, lang_code=self.lang_id.code)
+            )
+
+        return {
+            product: {
+                'id': product.default_code or product.id,
+                'title': product.with_context(display_default_code=False).display_name,
+                'description': product.website_meta_description or product.description_sale,
+                'link': format_product_link(product.website_url),
+                **self._prepare_gmc_identifier(product),
+                **self._prepare_gmc_image_links(product, base_url),
+                **price_info,
+                **self._prepare_gmc_stock_info(product),
+                **self._prepare_gmc_additional_info(product),
+            }
+            for product in products
+            if product._is_variant_possible()
+            and (price_info := self._prepare_gmc_price_info(product))
+        }
+
+    def _get_feed_product_domain(self):
+        product_domain = Domain.AND([
+            Domain('is_published', '=', True),
+            Domain('type', 'in', ('consu', 'combo')),
+            self.website_id.website_domain(),
+        ])
+        if self.product_category_ids:
+            product_domain &= Domain('public_categ_ids', 'child_of', self.product_category_ids.ids)
+
+        return product_domain
+
+    def _get_feed_products(self):
+        product_domain = self._get_feed_product_domain()
+
+        products = self.env['product.product'].search(
+            product_domain, limit=const.PRODUCT_FEED_HARD_LIMIT
+        )
+
+        # Send an early warning to the website manager if the number of products exceeds the
+        # midpoint between the soft and hard limit.
+        if len(products) > (const.PRODUCT_FEED_SOFT_LIMIT + const.PRODUCT_FEED_HARD_LIMIT) / 2:
+            self._notify_website_manager(
+                subject=self.env._("GMC: Product Limit Exceeded"),
+                body=self.env._(
+                    "The feed %(feed_name)s contains more than %(limit)s products,"
+                    " which may not be fully updated. Consider refining the feed by adjusting"
+                    " the product categories.",
+                    feed_name=self.display_name,
+                    limit=f"{const.PRODUCT_FEED_SOFT_LIMIT:,}",  # Format to 5,000
+                ),
+            )
```

Comment:
save on the feed whether manager was already notified to avoid spamming every hour/... if he's busy/on holiday :eyes:

Maybe at most once a week.